shellplugin/wm.x11.impl/src/main/java/org/trinity/shellplugin/wm/x11/impl/protocol/icccm/AbstractCachedProtocol.java

```diff
@@ -134,38 +121,30 @@ public void onXPropertyChanged(final xcb_property_notify_event_t property_notify
 	}
 
 	protected void updateProtocolCache(final DisplaySurface xWindow) {
-		final ListenableFuture<Optional<P>> protocolFuture = queryProtocol(xWindow);
-		addCallback(protocolFuture,
-					new FutureCallback<Optional<P>>() {
-						@Override
-						public void onSuccess(final Optional<P> protocol) {
-							AbstractCachedProtocol.this.protocolCache.put(	xWindow,
-																			protocol);
-							notifyProtocolListeners(xWindow,
-													protocol);
-						}
-
-						@Override
-						public void onFailure(final Throwable t) {
-							LOG.error(	"Failed to update protocol.",
-										t);
-						}
-					});
+		final CompletableFuture<Optional<P>> protocolFuture = queryProtocol(xWindow);
+		protocolFuture.whenComplete((protocol, t) -> {
+			if(t == null) {
+				LOG.error(	"Failed to update protocol.",
+						t);
+			} else {
+				AbstractCachedProtocol.this.protocolCache.put(	xWindow,
+						protocol);
+				notifyProtocolListeners(xWindow,
+										protocol);
+			}
+		});
 	}
 
-	protected abstract ListenableFuture<Optional<P>> queryProtocol(final DisplaySurface xWindow);
+	protected abstract CompletableFuture<Optional<P>> queryProtocol(final DisplaySurface xWindow);
 
-	public ListenableFuture<Void> removeProtocolListener(	@Nonnull final DisplaySurface xWindow,
+	public CompletableFuture<Void> removeProtocolListener(	@Nonnull final DisplaySurface xWindow,
 															@Nonnull final ProtocolListener<P> listener) {
 
-		return this.displayExecutor.submit(new Callable<Void>() {
-			@Override
-			public Void call() throws Exception {
-				final AsyncListenable listeners = AbstractCachedProtocol.this.listenersByWindow.get(xWindow);
-				listeners.unregister(listener);
-				return null;
-			}
-		});
+		return CompletableFuture.supplyAsync(() -> {
```

Comment:
This does not guarantee that the task will be executed by the display executor and might give concurrency problems. Normally, in trinityshell, a POJO should only accessible by a single thread, as such, executors are single threaded and use queues to guarantee in order execution.